/** ****************************************************************************
 * Главная форма приложения C&S.
 *
 * \file mainwindow.h
 * ****************************************************************************/

#ifndef MAINWINDOW_H
#define MAINWINDOW_H

#include <QMainWindow>
#include <QString>
#include <QFileDialog>

/** ****************************************************************************
 * Для работы с тэгами MP3
 * ****************************************************************************/
#include "ID3v2TagWorker.h"

/** ****************************************************************************
 * Для шифрования
 * ****************************************************************************/
#include "cryptor.h"

/** ****************************************************************************
 * \brief Режимы работы приложения
 * ****************************************************************************/
enum eModeWork
{
    READ,
    WRITE
};

QT_BEGIN_NAMESPACE
namespace Ui { class MainWindow; }
QT_END_NAMESPACE

class MainWindow : public QMainWindow
{
    Q_OBJECT

public:
    /** ************************************************************************
     * \brief Конструктор класса
     *
     * \param [in] parent Родительский объект
     * ************************************************************************/
    MainWindow(QWidget *parent = nullptr) noexcept;

    /** ************************************************************************
     * \brief Деструктор класса
     * ************************************************************************/
    ~MainWindow() noexcept;

private slots:
    /** ************************************************************************
     * \brief При смене текущего режима работы программы
     *
     * \param [in] Индекс режима работы
     * ************************************************************************/
    void on_cbProgrammMode_currentIndexChanged(int index);

    /** ************************************************************************
     * \brief Нажатие кнопки выбора файла
     * ************************************************************************/
    void on_btnChoseFile_clicked();

    /** ************************************************************************
     * \brief Нажатие кнопки старт
     * ************************************************************************/
    void on_btnStart_clicked();

private:
    /** ************************************************************************
     * \brief Для доступа к UI форме
     * ************************************************************************/
    Ui::MainWindow *ui;

    /** ************************************************************************
     * \brief Имя файла
     * ************************************************************************/
    QString fileName;

    /** ************************************************************************
     * \brief Сообщение
     * ************************************************************************/
    QString message;

    /** ************************************************************************
     * \brief Текущий режим работы
     * ************************************************************************/
    eModeWork currentModeWork;

    /** ************************************************************************
     * \brief Количество раундов для шифрования
     * ************************************************************************/
    const quint32 roundsCount = 22;

    /** ************************************************************************
     * \brief Вывод сообщения в статус бар
     *
     * \param [in] message Сообщение для вывода
     * ************************************************************************/
    void printToStatusBar(const QString &message) noexcept;

    /** ************************************************************************
     * \brief Обновить доступность кнопок и тд на UI
     * ************************************************************************/
    void updateUI() noexcept;

    /** ************************************************************************
     * \brief Считать сообщение
     * ************************************************************************/
    void readMessage();

    /** ************************************************************************
     * \brief Записать сообщение
     * ************************************************************************/
    void writeMessage();
};

#endif // MAINWINDOW_H
